# Assignment 08
## Matt Briskey

In [1]:
import pandas as pd

# the mpg data is used for demonstration
df = pd.read_csv("https://raw.githubusercontent.com/franklin-univ-data-science/data/master/mpg.csv")
df.head()

mpg  cylinders  displacement  acceleration  model  origin
0  18.0          8         307.0          12.0     70       1
1  15.0          8         350.0          11.5     70       1
2  18.0          8         318.0          11.0     70       1
3  16.0          8         304.0          12.0     70       1
4  17.0          8         302.0          10.5     70       1

In [2]:
df.shape

(406, 6)

In [3]:
# Create a new dataframe that shows only rows that have 4 cylinders 
df4 = df.loc[df['cylinders'] == 4]
df4

mpg  cylinders  displacement  acceleration  model  origin
10    NaN          4         133.0          17.5     70       2
20   24.0          4         113.0          15.0     70       3
24   27.0          4          97.0          14.5     70       3
25   26.0          4          97.0          20.5     70       2
26   25.0          4         110.0          17.5     70       2
..    ...        ...           ...           ...    ...     ...
401  27.0          4         140.0          15.6     82       1
402  44.0          4          97.0          24.6     82       2
403  32.0          4         135.0          11.6     82       1
404  28.0          4         120.0          18.6     82       1
405  31.0          4         119.0          19.4     82       1

[207 rows x 6 columns]

In [4]:
df4.shape

(207, 6)

In [5]:
def summarize_data(g):
  return(
    pd.Series({
      "avg_MPG": g.mpg.mean(),
      "min_MPG": g.mpg.min(),
      "max_MPG": g.mpg.max(),
      "sample_count": len(g),
      "mix": g.cylinders.sum()/df.cylinders.sum(),
    })
  )

df.groupby("cylinders").apply(summarize_data)

avg_MPG  min_MPG  max_MPG  sample_count       mix
cylinders                                                     
3          20.550000     18.0     23.7           4.0  0.005398
4          29.286765     18.0     46.6         207.0  0.372470
5          27.366667     20.3     36.4           3.0  0.006748
6          19.985714     15.0     38.0          84.0  0.226721
8          14.963107      9.0     26.6         108.0  0.388664

### Read the summarize_data() and explain the difference between g and df.

df is the dataframe while g is the aggregated data grouped by number of cylinders.  df has 406 different rows of data and g groups all of the data by number of cylinders to get mean, min, max, count, and % of total for each grouped number of cylinders.  For example, there are 207 rows of data in df that have 4 cylinders.  g groups all of these 207 rows into an aggregrate group and then calculates the mean, min, max, count, and % of total for that specific group.

In [6]:
# Select the measure and dimention variables from your dataframe, apply the summarize_data() 
# and modify as needed based on your preferences.

df2 = pd.read_csv("https://raw.githubusercontent.com/MattBriskey/data612_data_mining/main/data_sets/Spotify%20Streams.csv")
df2 = df2.head(75) #Take the first 75 rows of the dataset

In [7]:
# List the column names
for col in df2.columns:
    print(col)

Song
Artist
Streams_Billions
Release Date


In [8]:
# List the data type for each column
df2.dtypes

Song                 object
Artist               object
Streams_Billions    float64
Release Date         object
dtype: object

In [15]:
# Custom function to return the avg, min, max, count, and % of total grouped by Artist
def summarize_data(g):
  return(
    pd.Series({
      "avg_streams": g.Streams_Billions.mean(),
      "min_streams": g.Streams_Billions.min(),
      "max_streams": g.Streams_Billions.max(),
      "sample_count": len(g),
      "mix": g.Artist.count()/df2.Artist.count(),
    })
  )

df3 = df2.groupby('Artist').apply(summarize_data)
df5 = df3.sort_values(by='sample_count', ascending=False) #Sort the resulting dataframe in descending order by mix

format_dict = {
    'avg_streams':'{0:,.2f}', 
    'min_streams': '{0:,.2f}',  
    'max_streams': '{0:,.2f}',
    'sample_count': '{0:,.0f}',
    'mix': '{:.2%}'}
df5.style.format(format_dict).highlight_max(color = 'lightgreen', axis = 0) # highlights the max value of each column in green
